### PR TITLE
Cleanup reading page index (#4149) (#4090)

### LINKED ISSUE #4149: Parquet Page Index Reader Assumes Consecutive Offsets
**Describe the bug**

The parquet page index reading logic assumes that the index is contiguous, and in the case of the offset index it assumes it is in order.

The [specification](https://github.com/apache/parquet-format/blob/master/PageIndex.md#technical-approach) strongly suggests this is a valid assumption, however, it never explicitly specifies it. We should probably be defensive and assume that there is a writer out there that doesn't do this.

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4149
Precursor to #4090

# Rationale for this change
 

This performs some cleanup on the logic for reading the page index, in preparation for adding an async API. It also eliminates the assumption of contiguous ranges (#4149).

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/file/page_index/index_reader.rs [hunk 1]
```diff
@@ -24,9 +24,21 @@ use crate::file::metadata::ColumnChunkMetaData;
 use crate::file::page_index::index::{Index, NativeIndex};
 use crate::file::reader::ChunkReader;
 use crate::format::{ColumnIndex, OffsetIndex, PageLocation};
-use std::io::{Cursor, Read};
+use std::io::Cursor;
+use std::ops::Range;
 use thrift::protocol::{TCompactInputProtocol, TSerializable};
 
+/// Computes the aggregate range of two optional ranges
```
Comment (crepererum) [inline parquet/src/file/page_index/index_reader.rs:31]: What's an "aggregate range"?

**Review by crepererum — COMMENTED** [review]
Code looks cleaner.

Two nitpicks:

1. docstring (see comment)
2. can we have a regression test for #4149?

File: parquet/src/file/page_index/index_reader.rs (hunk 1 above)
Comment (tustvold) [reply parquet/src/file/page_index/index_reader.rs]: Tried to clarify in https://github.com/apache/arrow-rs/pull/4151/commits/3809a4d33e704536e41fd62d8656b6ed0afe7560

File: parquet/src/file/metadata.rs [hunk 2]
```diff
@@ -565,6 +566,13 @@ impl ColumnChunkMetaData {
         self.column_index_length
     }
 
+    /// Returns the range for the offset index if any
+    pub(crate) fn column_index_range(&self) -> Option<Range<usize>> {
```
Comment (alamb) [inline parquet/src/file/metadata.rs:570]: Maybe we should make these functions pubic and deprecate the `offset_index_offset()` and `offset_index_length()` `column_index_offset()` and `column_index_length()` functions 🤔 `

File: parquet/src/arrow/async_reader/mod.rs [hunk 3]
```diff
@@ -240,78 +243,53 @@ impl<T: AsyncFileReader + Send + 'static> ArrowReaderBuilder<AsyncReader<T>> {
...
+                None,
+                |a, c| {
+                    let a = acc_range(a, c.column_index_range());
+                    acc_range(a, c.offset_index_range())
+                },
+            );
+
+            if let Some(fetch) = fetch {
+                let bytes = input.get_bytes(fetch.clone()).await?;
+                let bytes = |r: Range<usize>| {
```
Comment (alamb) [inline parquet/src/arrow/async_reader/mod.rs:256]: I was confused at first as the `bytes` closure shadowed the `bytes` actual data

Maybe calling the closure something like `get_bytes` or `bytes_from_range` would be more readable 🤔 

This same pattern appears several more times below in this PR

**Review by alamb — APPROVED** [review]
Looks like a very nice cleanup to me. Thank you @tustvold 

I also verified the test coverage. Without the changes in this PR the test fails like this:


```
---- file::serialized_reader::tests::test_page_index_reader_out_of_order stdout ----
thread 'file::serialized_reader::tests::test_page_index_reader_out_of_order' panicked at 'called `Result::unwrap()` on an `Err` value: External(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })', parquet/src/file/serialized_reader.rs:1400:65


```

File: parquet/src/file/metadata.rs (hunk 2 above)
Comment (tustvold) [reply parquet/src/file/metadata.rs:570]: Definitely something to consider as I iterate on these APIs

File: parquet/src/arrow/async_reader/mod.rs (hunk 3 above)
Comment (alamb) [reply parquet/src/arrow/async_reader/mod.rs]: `get` looks better - thank you

**Pull request merged**